database/engine/journalfile.c

```diff
@@ -524,6 +521,614 @@ static uint64_t iterate_transactions(struct rrdengine_instance *ctx, struct rrde
+            if (!rc) {
+                total_pages += metric_list_header->entries;
+                verified++;
+            }
+        }
+
+        metric++;
+        if (((uint8_t *) metric - (uint8_t *) data_start) > (uint32_t) file_size) {
+            info("DBENGINE: verification failed EOF reached -- total entries %u, verified %u", entries, verified);
+            return 1;
+        }
+    }
+
+    if (entries != verified) {
+        info("DBENGINE: verification failed -- total entries %u, verified %u", entries, verified);
+        return 1;
+    }
+    info("DBENGINE: verification succeeded -- total entries %u, verified %u (%u total pages)", entries, verified, total_pages);
+
+    return 0;
+}
+
+int load_journal_file_v2(struct rrdengine_instance *ctx, struct rrdengine_journalfile *journalfile, struct rrdengine_datafile *datafile)
+{
+    int ret, fd;
+    uint64_t file_size;
+    char path[RRDENG_PATH_MAX];
+    struct stat statbuf;
+    uint32_t original_file_size = 0;
+
+    generate_journalfilepath(datafile, path, sizeof(path));
+    ret = stat(path, &statbuf);
+    if (!ret)
+        original_file_size = (uint32_t)statbuf.st_size;
+
+    generate_journalfilepath_v2(datafile, path, sizeof(path));
+
+    fd = open(path, O_RDONLY);
+    if (fd < 0) {
+        if (errno == ENOENT)
+            return 1;
+        ++ctx->stats.fs_errors;
+        rrd_stat_atomic_add(&global_fs_errors, 1);
+        error("DBENGINE: failed to open '%s'", path);
+        return 1;
+    }
+
+    ret = fstat(fd, &statbuf);
+    if (ret) {
+        error("DBENGINE: failed to get file information for '%s'", path);
+        close(fd);
+        return 1;
+    }
+
+    file_size = (size_t)statbuf.st_size;
+
+    if (file_size < sizeof(struct journal_v2_header)) {
+        error_report("Invalid file %s. Not the expected size", path);
+        close(fd);
+        return 1;
+    }
+
+    usec_t start_loading = now_realtime_usec();
+    uint8_t *data_start = mmap(NULL, file_size, PROT_READ, MAP_SHARED, fd, 0);
+    if (data_start == MAP_FAILED) {
+        close(fd);
+        return 1;
+    }
+    close(fd);
+
+    info("DBENGINE: checking integrity of '%s'", path);
+    int rc = check_journal_v2_file(data_start, file_size, original_file_size);
+    if (unlikely(rc)) {
+        if (rc == 2)
+            error_report("File %s needs to be rebuilt", path);
+        else if (rc == 3)
+            error_report("File %s will be skipped", path);
+        else
+            error_report("File %s is invalid and it will be rebuilt", path);
+
+        if (unlikely(munmap(data_start, file_size)))
+            error("DBENGINE: failed to unmap '%s'", path);
+
+        return rc;
+    }
+
+    struct journal_v2_header *j2_header = (void *) data_start;
+    uint32_t entries = j2_header->metric_count;
+
+    if (unlikely(!entries)) {
+        if (unlikely(munmap(data_start, file_size)))
+            error("DBENGINE: failed to unmap '%s'", path);
+
+        return 1;
+    }
+
```

Comment:
On `FreeBSD ` we cannot compile this PR due these errors:

```sh
[/home/thiago/Netdata/netdata]# gmake -j2 
database/engine/journalfile.c: In function 'load_journal_file_v2':
database/engine/journalfile.c:741:41: error: 'MADV_DONTFORK' undeclared (first use in this function); did you mean 'MADV_DONTNEED'?
  741 |     rc = madvise(data_start, file_size, MADV_DONTFORK);
      |                                         ^~~~~~~~~~~~~
      |                                         MADV_DONTNEED
database/engine/journalfile.c:741:41: note: each undeclared identifier is reported only once for each function it appears in
database/engine/journalfile.c:745:41: error: 'MADV_DONTDUMP' undeclared (first use in this function); did you mean 'MADV_DONTNEED'?
  745 |     rc = madvise(data_start, file_size, MADV_DONTDUMP);
      |                                         ^~~~~~~~~~~~~
      |                                         MADV_DONTNEED
gmake[2]: *** [Makefile:6159: database/engine/journalfile.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[1]: *** [Makefile:7398: all-recursive] Error 1
gmake: *** [Makefile:3761: all] Error 2
 FAILED  ''
```

we already addressed an issue in another PR due an error like this.

It is also possible that this has impact on Mac.

Reply:
I think I fixed these.

Reply:
I am waiting for final report to confirm this.

Reply:
Sorry I confused, you are right, you fixed.